### ISSUE TITLE
Unable to change field type because document's saveError is created in init()

### ISSUE THREAD

**Issue by napalm272** [Automattic/mongoose#3117]
I have a model which has a field called `state`.  `state` used to be a boolean, I want to change it to a number with a default value.  Since I already have some documents in my db, I tried to make a simple "migration" script, here's my code:
```
MyModel.find({})
  .exec(function(err, docs){
    async.each(docs,
      function(doc, cb){
        doc.set('state', 0);
        doc.save(cb);
...
```
However I get this error during the save : `CastError: Cast to number failed for value "true" at path "state"`
After digging around in mongoose's source code, I found this in document.js

```
function init (self, obj, doc, prefix) {
  prefix = prefix || '';

  var keys = Object.keys(obj)
    , len = keys.length
    , schema
    , path
    , i;

  while (len--) {
    i = keys[len];
    path = prefix + i;
    schema = self.schema.path(path);

    if (!schema && utils.isObject(obj[i]) &&
        (!obj[i].constructor || 'Object' == obj[i].constructor.name)) {
      // assume nested object
      if (!doc[i]) doc[i] = {};
      init(self, obj[i], doc[i], path + '.');
    } else {
      if (obj[i] === null) {
        doc[i] = null;
      } else if (obj[i] !== undefined) {
        if (schema) {
          self.$__try(function(){
            doc[i] = schema.cast(obj[i], self, true);
          });
        } else {
          doc[i] = obj[i];
        }
      }
      // mark as hydrated
      if (!self.isModified(path)) {
        self.$__.activePaths.init(path);
      }
    }
  }
};
```
It seems that the CastError is created by 
```
self.$__try(function(){
            doc[i] = schema.cast(obj[i], self, true);
          });
```
and then stored as $__.saveError, which causes the save to fail.

My question is why is this error created in init()?  Why not just check if the field has the correct type right before saving?  And how can I get around that if I want my migration script to work? (right now I'm manually setting $__.saveError to null before I save the document, which seems really hacky)

Thanks!

**Comment by napalm272** [comment]
Version is 3.8.31

**Comment by vkarpov15** [comment]
Upgrade to 4.0.x, this was how 3.8.x handled cast errors. Duplicate of #2237 etc. Fix won't be backported to 3.8.x because this is potentially backwards-breaking for cases like like #2997. No good way to work around this in 3.8.x other than clearing out `$__.saveError` unfortunately.
